lib/mainloop-io-worker.h

```diff
@@ -32,6 +32,7 @@ typedef struct _MainLoopIOWorkerJob
 {
   void (*work)(gpointer user_data);
   void (*completion)(gpointer user_data);
+  void (*destroy)(gpointer user_data);
   gpointer user_data;
```

Comment:
I might suggest renaming the destroy() function, it does not destroy the Job object at all. I thinkit merely cleans it up, so cleanup() would be more appropriate.

Also, I am a bit unconvinced about the coupling between the destroy callback and the fact that we are taking the ref outside of the responsibilities of the Job class. What if we had two methods? One that would take the ref (in the main thread) and another that would release it (in complete). Both would be methods of the MainLoopIOWorkerJob
class, thus the ref/unref dance would nicely be encapsulated and would be easy to follow.

The tricky question is how we would name these functions. What about:
  * engage(): right before submitting the job
  * release(): at completion

This way we wouldn't even need to call destroy in the main_loop_io_worker_submit() function, as we would bail before calling engage().

What do you think?

Reply:
I like the preparation+cleanup function idea, and the engage/release names also sound good. I will rewrite the patch according to your suggestion.